op-proposer/proposer/driver.go

```diff
@@ -471,27 +490,48 @@ func (l *L2OutputSubmitter) loopL2OO(ctx context.Context) {
 func (l *L2OutputSubmitter) loopDGF(ctx context.Context) {
 	ticker := time.NewTicker(l.Cfg.ProposalInterval)
 	defer ticker.Stop()
+
+	retryOutputProposal := false
+	retryTicker := time.NewTicker(l.Cfg.ProposalRetryInterval)
+	defer retryTicker.Stop()
+
+	proposeOutput := func() bool {
+		blockNumber, err := l.FetchCurrentBlockNumber(ctx)
+		if err != nil {
+			retryTicker.Reset(l.Cfg.ProposalRetryInterval)
+			return true
+		}
+
+		output, shouldPropose, err := l.FetchOutput(ctx, blockNumber)
+		if err != nil || !shouldPropose {
+			retryTicker.Reset(l.Cfg.ProposalRetryInterval)
+			return true
+		}
+
+		err = l.proposeOutput(ctx, output)
+		if err != nil {
+			retryTicker.Reset(l.Cfg.ProposalRetryInterval)
+			return true
+		}
+		return false
+	}
+
 	for {
 		select {
 		case <-ticker.C:
-			blockNumber, err := l.FetchCurrentBlockNumber(ctx)
-			if err != nil {
-				break
-			}
-
-			output, shouldPropose, err := l.FetchOutput(ctx, blockNumber)
-			if err != nil || !shouldPropose {
-				break
+			retryOutputProposal = proposeOutput()
+		case <-retryTicker.C:
+			if retryOutputProposal {
+				l.Log.Info("retrying output proposal")
+				retryOutputProposal = proposeOutput()
```

Comment:
I need to find the best spot to test this code. There aren't currently many tests in op-proposer, so leaning towards adding a op-e2e test. Open to suggestions though!

Reply:
This approach seems kind of complex since it depends on behavior of `select`. Wouldn't it be better to wrap our retry helpers around functions like this one? https://github.com/ethereum-optimism/optimism/blob/59fbf91ec45f0c636d37b4c7bbab3bb7a1c79440/op-proposer/proposer/driver.go#L281

That would prevent the error from bubbling up in the first place

Reply:
I agree, I think a simpler solution is really to just wrap the `l.FetchOutput(ctx, blockNumber)` call in `retry.Do2` call with a `retry.Fixed` strategy, and otherwise restore the old code. Applies to both, L2OO and DGF.

The `l.proposeOutput(ctx, output)` call shouldn't be part of the retry, as retrying failing txs is already handled within the txmgr.

Reply:
As for the test, you could write two unit tests for the two loops `loopL2OO` and `loopDGF` using a mock `RollupProvider` that returns a mock `RollupClient` and then set the mock up to return a couple of errors before returning the right value (and also set the mock up to return a sync status). There are other tests that set these mocks up already, and the generated mocks already exist.

You would also need to set up the `Txmgr` mock and assert that it receives a correct `Send` call and a bunch of other getter calls.

These are two non-trivial tests to add, but a nice bonus is that we'd finally have unit tests of the proposer loops, which are currently missing.

Reply:
sorry for the confusion @bitwiseguy we totally should also include `l.FetchCurrentBlockNumber(ctx)` in the retrying.

Reply:
Oh I see the issue with mocking the contract calls. Maybe what we do is just use an interface for the `l2ooContract` field, we only access the two methods `NextBlockNumber` and `Version`. You can then set that field in the test to a mock. Sorry for the mocking mess.

Reply:
I think the test can just create the proposer struct directly and only test the loops, so we don't need use any of the `new*Submitter` functions.

Reply:
Yea good idea to use interface for `l2ooContract`. I will try that.

For the test is it acceptable to trigger the `loopL2OO` and then just set a sleep/timer long enough that we know some retries are triggered, then check that mocked functions were called as expected? Generally I try to avoid sleep/timer in tests but not sure if there is a better way in this case since `loopL2OO` runs a forever loop

Reply:
We can set the proposer and retry intervals super short in the test, maybe even to 0. Then the mock could be set up so that the n'th time `SendTransaction` is called, it closes the `l.done` channel and the loop should return.

Reply:
Not sure if it helps in this case, but we do have [a clock abstraction](https://github.com/ethereum-optimism/optimism/blob/770d37f86fccec38596a134bf9bbe075f08ebd10/op-service/clock/clock.go#L12) in `op-service/clock` which has a test stub so you can deterministically control time, including stepping forward a certain amount of time and having any created timers trigger deterministically.